### ISSUE TITLE
[BUG] [TGL Chrome] Fw panic when we overwrite the firmware & topology binary to the file system first time and boot it

### ISSUE THREAD

**Issue by RDharageswari** [thesofproject/sof#4419]
**Describe the bug**
Fw panic when we copy the firmware & topology binary to the file system first time and boot it

**To Reproduce**
Steps to reproduce the behavior: (e.g. list commands or actions used to reproduce the bug)

**Reproduction Rate**
whenever we copy the firmware & topology for the first time on a board.
The problem goes away if we reboot

**Expected behavior**
No firmware panic and the sound card should bind successfully

**Impact**
showstopper

**Environment**
tgl-13 release.

**Screenshots or console output**
[    7.550234] sof-audio-pci 0000:00:1f.3: error : DSP panic!
[    7.550243] sof-audio-pci 0000:00:1f.3: status: fw entered - code 00000005
[    7.550456] sof-audio-pci 0000:00:1f.3: error: runtime exception
[    7.550458] sof-audio-pci 0000:00:1f.3: error: trace point 00004000
[    7.550459] sof-audio-pci 0000:00:1f.3: error: panic at :0
[    7.550461] sof-audio-pci 0000:00:1f.3: error: DSP Firmware Oops
[    7.550463] sof-audio-pci 0000:00:1f.3: error: Exception Cause: LoadStorePIFD                                                                             ataErrorCause, Synchronous PIF data error during LoadStore access
[    7.550465] sof-audio-pci 0000:00:1f.3: EXCCAUSE 0x0000000d EXCVADDR 0xbe1d00                                                                             00 PS       0x00060925 SAR     0x00000000
[    7.550466] sof-audio-pci 0000:00:1f.3: EPC1     0xbe02e3b7 EPC2     0xbe0424                                                                             0a EPC3     0x00000000 EPC4    0x00000000
[    7.550468] sof-audio-pci 0000:00:1f.3: EPC5     0x00000000 EPC6     0x000000                                                                             00 EPC7     0x00000000 DEPC    0x00000000
[    7.550470] sof-audio-pci 0000:00:1f.3: EPS2     0x00060b20 EPS3     0x000000                                                                             00 EPS4     0x00000000 EPS5    0x00000000
[    7.550471] sof-audio-pci 0000:00:1f.3: EPS6     0x00000000 EPS7     0x000000                                                                             00 INTENABL 0x00000000 INTERRU 0x00000222
[    7.550473] sof-audio-pci 0000:00:1f.3: stack dump from 0xbe131360
[    7.550476] sof-audio-pci 0000:00:1f.3: 0xbe131360: 3fe6b541 a0000000 3fe5ad5                                                                             0 3eb5aa0d
[    7.550478] sof-audio-pci 0000:00:1f.3: 0xbe131364: 00000000 3ff99999 9980000                                                                             0 be136fb0
[    7.550479] sof-audio-pci 0000:00:1f.3: 0xbe131368: 00000000 00000000 0000000                                                                             0 00000000
[    7.550481] sof-audio-pci 0000:00:1f.3: 0xbe13136c: b66db3a0 ffff89ff 45573dc                                                                             0 ffffb2e5
[    7.550483] sof-audio-pci 0000:00:1f.3: 0xbe131370: 8ec1bb40 ffffffff 8f21b70                                                                             0 ffffffff
[    7.550484] sof-audio-pci 0000:00:1f.3: 0xbe131374: 5c59e028 ffff89ff 000c080                                                                             0 00000000
[    7.550486] sof-audio-pci 0000:00:1f.3: 0xbe131378: 00000000 00000000 45573e5                                                                             0 ffffb2e5
[    7.550488] sof-audio-pci 0000:00:1f.3: 0xbe13137c: 45573e78 ffffb2e5 c055216                                                                             3 ffffffff
[    7.812424] NET: Registered protocol family 38
[    7.878571] x86/PAT: flashrom:2707 conflicting memory types fe000000-10000000                                                                             0 uncached-minus<->uncached
[    7.878575] x86/PAT: reserve_memtype failed [mem 0xfe000000-0xffffffff], trac                                                                             k uncached-minus, req uncached-minus
[    7.889983] Bluetooth: rfcomm_init_sockets() RFCOMM socket layer initialized
[    7.890009] Bluetooth: rfcomm_init() RFCOMM ver 1.11
[    7.908778] NET: Registered protocol family 40
[    7.932593] Bluetooth: msft_le_set_advertisement_filter_enable_cb() hci0: MSF                                                                             T filter_enable is already on
[    7.957788] input: cros_fp_input as /devices/virtual/input/input13
[    8.051731] sof-audio-pci 0000:00:1f.3: error: ipc timed out for 0x40070000 s                                                                             ize 12
[    8.051740] sof-audio-pci 0000:00:1f.3: status: fw entered - code 00000005
[    8.051947] sof-audio-pci 0000:00:1f.3: error: runtime exception
[    8.051948] sof-audio-pci 0000:00:1f.3: error: trace point 00004000
[    8.051950] sof-audio-pci 0000:00:1f.3: error: panic at :0
[    8.051951] sof-audio-pci 0000:00:1f.3: error: DSP Firmware Oops
[    8.051953] sof-audio-pci 0000:00:1f.3: error: Exception Cause: LoadStorePIFD                                                                             ataErrorCause, Synchronous PIF data error during LoadStore access
[    8.051954] sof-audio-pci 0000:00:1f.3: EXCCAUSE 0x0000000d EXCVADDR 0xbe1d00                                                                             00 PS       0x00060925 SAR     0x00000000
[    8.051956] sof-audio-pci 0000:00:1f.3: EPC1     0xbe02e3b7 EPC2     0xbe0424                                                                             0a EPC3     0x00000000 EPC4    0x00000000
[    8.051957] sof-audio-pci 0000:00:1f.3: EPC5     0x00000000 EPC6     0x000000                                                                             00 EPC7     0x00000000 DEPC    0x00000000
[    8.051965] sof-audio-pci 0000:00:1f.3: EPS2     0x00060b20 EPS3     0x000000                                                                             00 EPS4     0x00000000 EPS5    0x00000000
[    8.051966] sof-audio-pci 0000:00:1f.3: EPS6     0x00000000 EPS7     0x000000                                                                             00 INTENABL 0x00000000 INTERRU 0x00000222
[    8.051968] sof-audio-pci 0000:00:1f.3: stack dump from 0xbe131360
[    8.051970] sof-audio-pci 0000:00:1f.3: 0xbe131360: 3fe6b541 a0000000 3fe5ad5                                                                             0 3eb5aa0d
[    8.051971] sof-audio-pci 0000:00:1f.3: 0xbe131364: 00000000 3ff99999 9980000                                                                             0 be136fb0
[    8.051972] sof-audio-pci 0000:00:1f.3: 0xbe131368: 00000000 00000000 0000000                                                                             0 00000000
[    8.051979] sof-audio-pci 0000:00:1f.3: 0xbe13136c: 406535c0 ffffb2e5 0000328                                                                             2 00000000
[    8.051980] sof-audio-pci 0000:00:1f.3: 0xbe131370: 406535a0 ffffb2e5 2883140                                                                             0 49f1c1f5
[    8.051985] sof-audio-pci 0000:00:1f.3: 0xbe131374: b49380d8 ffff89ff 5c59e02                                                                             8 ffff89ff
[    8.051986] sof-audio-pci 0000:00:1f.3: 0xbe131378: 00000000 00000000 406536f                                                                             0 ffffb2e5
[    8.051987] sof-audio-pci 0000:00:1f.3: 0xbe13137c: b4938000 ffff89ff 8dd6932                                                                             5 ffffffff
[    8.052014] sof-audio-pci 0000:00:1f.3: error: hda irq intsts 0x00000000 intl                                                                             ctl 0xc0000000 rirb 00
[    8.052015] sof-audio-pci 0000:00:1f.3: error: dsp irq ppsts 0x00000000 adspi                                                                             s 0x00000000
[    8.052033] sof-audio-pci 0000:00:1f.3: error: host status 0x00000000 dsp sta                                                                             tus 0x00000000 mask 0x00000003
[    8.052035] sof-audio-pci 0000:00:1f.3: error: core 1 enable ipc failure -110
[    8.052040] sof-audio-pci 0000:00:1f.3: error: enable core: -110
[    8.052055] sof-audio-pci 0000:00:1f.3: ASoC: failed to load widget PCM99C

**Comment by RDharageswari** [comment]
[FW_panic_while_boot.txt](https://github.com/thesofproject/sof/files/6729194/FW_panic_while_boot.txt)

**Comment by lgirdwood** [comment]
@RDharageswari can you confirm that you copy files (how ? over NFS, scp ?) and then immediately reboot (how) ?
This could also be kernel issue or FS cache. Have you tried copy files then unload/reload modules ?
@ujfalusi anything in kernel flow for FW loading that would break here ?

**Comment by bkokoszx** [comment]
@lgirdwood @RDharageswari 
I was able to reproduce this issue. I remove tplg and fw files, then I copy tplg/fw using scp and reboot by ```sudo reboot```. Sometimes after reboot I get dsp panic in dmesg. 

Once I also had a case, that just after reboot I did not have panic, but then I tried to run: 
```
aplay -Dhw:0,0 tmp.wav -c2 -r48000 -fs24_le -v
```
and after that I got similar panic.

**Comment by ujfalusi** [comment]
Can it be not related to firmware copying and it is just a coincident? Or is it more likely to happen after delete fw/tplg, copy new fw/tplg, then reboot?
Fwiw, in the log the panic happens when loading the topology, specifically with a timeout of sending (SOF_IPC_GLB_PM_MSG|SOF_IPC_PM_CORE_ENABLE) message (`ipc timed out for 0x40070000 size 12`), it is hard to say if it is for the `pipeline->core` or for the `swidget->core` and also from where it happens.

Can it be reproduced if _only_ the firmware or the tplg file is deleted/copyied ?

After the first DSP crash you can reboot repeatably without any issue?

Can it be caused by unclean shutdown of the DSP in a way that same stale data remain from the previous boot and when we load the new tplg, the firmware tries to interpret it and fails?

Can you try to delete the fw/tplg files and then copy the same ones back? Does that also triggers a crash?
Can you try to update the fw/tplg and then shut down and then power on? Does the crash remain?
Can you reboot repeatedly without touching the fw/tplg and see if there is really no DSP crash?

I have been doing locally: update the fw/tplg, unload all drivers, load the drivers again and now the new firmware is in use. Can you try the same, so update the fw/tplg then remove/insert the modules and see if the crash happens also like that?

**Comment by ranj063** [comment]
@RDharageswari can you confirm this is tgl-013 specific problem?

**Comment by RDharageswari** [comment]
I have seen this problem only with tgl-13. Did not hit it before.

**Comment by plbossart** [comment]
This reminds me of the odd behavior with firmware being changed reported by @marc-hb https://github.com/thesofproject/linux/pull/2954

Some sort of firmware caching issue?

**Comment by marc-hb** [comment]
I'm confused by the description of this bug: is the same, valid and sane firmware used all the time in every single reproduction steps?

The problem solved (for me) by https://github.com/thesofproject/linux/pull/2954 is much simpler to describe:
1. Load a firmware that invokes panic() _voluntarily at a certain point._
2. That panicky firmware gets CACHED until the next reboot, no way to replace it. The kernel drivers keep reloading the same version, it's stuck to it no matter what's in `/lib/firmware/`, including nothing at all.

I'm resuming the investigation on 2954 now in case it helps this too.

PS: why is this P1? It affects only developers and no product, correct?

**Comment by ranj063** [comment]
> PS: why is this P1? It affects only developers and no product, correct?

no the issue is seen on a TGL product

**Comment by marc-hb** [comment]
I meant: this does not seem like it would affect a _consumer_, sorry for the confusion.

**Comment by lgirdwood** [comment]
> Some sort of firmware caching issue?

This is what I was thinking too, VFS cache could do this but ```sudo reboot``` done by @bkokoszx should do a ```sync``` and writeback to disk.

@bkokoszx can you try and execute ```sync fw_file_name``` to forcibly flush VFS cache version of FW prior to reboot.

**Comment by bkokoszx** [comment]
@lgirdwood 
I did so and I was able to reproduce this issue with ```sync fw_file``` before ```sudo reboot```.

I was also trying to analyze exception and stack logs from dmesg with ```sof-tgl.lst``` file. I do not have any special conclusions yet, but we can see that:
```
EXCVADDR 0xbe1d0000
```
points to the:
```
be1d0000 <_sof_core_s_start>
```
, which is used in: ```cavs_pm_runtime_core_en_memory()``` and ```cavs_pm_runtime_core_dis_memory()``` functions. I will try to debug more.

**Comment by ranj063** [comment]
@bkokoszx https://github.com/thesofproject/linux/pull/3018 fixes the issue. I still need to refine it to align with other multi-core changes but the underlying problem is to disables the secondary cores before shutdown

**Comment by marc-hb** [comment]
According to the bug description the problem happens when the host boots. How can a DSP shutdown fix make a difference?

**Comment by ranj063** [comment]
its not just a boot, its a reboot. So shutdown can make a difference. We've seen issues similar to this in the past and we have a shutdown callback for the PCI device to put the DSP in D3 already. This just adds on to it

**Comment by RDharageswari** [comment]
Today i was still able to repro the issue, even with https://github.com/thesofproject/linux/pull/3018

**Comment by RDharageswari** [comment]
Even when the same firmware file is overwritten we hit the issue.
This issue is not consistently observed every time.

**Comment by RDharageswari** [comment]
with the following changes @bkokoszx suggested:
diff --git a/src/platform/intel/cavs/lib/pm_runtime.c b/src/platform/intel/cavs/lib/pm_runtime.c
index aba17421135a..84970c6366fc 100644
--- a/src/platform/intel/cavs/lib/pm_runtime.c
+++ b/src/platform/intel/cavs/lib/pm_runtime.c
@@ -328,8 +328,8 @@ static inline void cavs_pm_runtime_core_dis_memory(uint32_t index)
        /* Address is calculated for index (0 for the primary core) minus one
         * since _sof_core_s_start is first secondary core stack address
         */
-       core_memory_ptr = (char *)&_sof_core_s_start
-               + (index - 1) * SOF_CORE_S_SIZE;
+       core_memory_ptr = cache_to_uncache((char *)&_sof_core_s_start
+               + (index - 1) * SOF_CORE_S_SIZE);

        cavs_pm_memory_hp_sram_power_gate(core_memory_ptr, SOF_CORE_S_SIZE,
                                          false);
@@ -346,8 +346,8 @@ static inline void cavs_pm_runtime_core_en_memory(uint32_t index)
        /* Address is calculated for index (0 for the primary core) minus one
         * since _sof_core_s_start is first secondary core stack address
         */
-       core_memory_ptr = (char *)&_sof_core_s_start
-               + (index - 1) * SOF_CORE_S_SIZE;
+       core_memory_ptr = cache_to_uncache((char *)&_sof_core_s_start
+               + (index - 1) * SOF_CORE_S_SIZE);

        cavs_pm_memory_hp_sram_power_gate(core_memory_ptr, SOF_CORE_S_SIZE,
                                          true);
I am not hitting the failure...tried multiple times..will ocntinue my testing

**Comment by marc-hb** [comment]
Additional comments in #4441

**Comment by marc-hb** [comment]
I just got to the bottom of the other issue: the failure to reload a new firmware after a firmware panic, see latest commit message in `release_firmware()` fix https://github.com/thesofproject/linux/pull/2954

tl;dr: it was not "caching" but "batching"

This could interact with this different issue and make it even harder to test, so I suggest testing with https://github.com/thesofproject/linux/pull/2954 too.

**Comment by bkokoszx** [comment]
I've also tried change mentioned by @RDharageswari (https://github.com/thesofproject/sof/issues/4419#issuecomment-872588622) and I did not have reproduction (tried several times ~20).

**Comment by slawblauciak** [comment]
Relevant PR created here: https://github.com/thesofproject/sof/pull/4446

**Comment by slawblauciak** [comment]
The PR is unfortunately not correct as it prevents SRAM power gating. However, that does point at a probable area where the fault exists. There appears to be a race condition happening with SRAM power gating. Simply adding as much as one additional trace point for debug makes the issue go away, making this problem extremely difficult to debug. However, that does present a potential workaround: simply delaying the power gating mechanism very slightly.
Unfortunately I have no other conclusions on this matter at this point, I'm still debugging it.

**Comment by slawblauciak** [comment]
@RDharageswari please try this PR https://github.com/thesofproject/sof/pull/4466

It makes the delay for power gating HP SRAM longer.

**Comment by sathya-nujella** [comment]
> @RDharageswari please try this PR #4466
> 
> It makes the delay for power gating HP SRAM longer.

Hi @slawblauciak 
FYI,
I got a script which does the below:
chrome DUT:
       loop {
	mount -o remount,rw /
	cp /root/sof-tgl.ri /lib/firmware/intel/sof/community/
	cp /root/sof-tgl-max98373-rt5682.tplg /lib/firmware/intel/sof-tplg/voxel/
         sync
        reboot 
       }
Within 10 cycles of the script, I was able to repro the original issue with tgl_13.

   Using this script, I have verified PR #4466 : 'DSP panic' issue is **not reproducible**.  Thanks!

**Comment by marc-hb** [comment]
@sathya-nujella I really don't understand how copying the firmware files can make any difference. Linux does not have any `/lib/firmware` "memory" that persists across reboots, if it had that would be a serious (security?) issue.

Can you please try to reproduce with the same loop but without the copies? (and without #4466). Everything else the same.

In other words I don't believe the title of this bug.

**Comment by sathya-nujella** [comment]
> @sathya-nujella I really don't understand how copying the firmware files can make any difference. Linux does not have any `/lib/firmware` "memory" that persists across reboots, if it had that would be a serious (security?) issue.
> 
> Can you please try to reproduce with the same loop but without the copies? (and without #4466). Everything else the same.
> 
> In other words I don't believe the title of this bug.

Yes @marc-hb . 
Validation teams has tried normal reboot cycles without copying any firmware & topologies: no issues reported.

Issue did not report in previous TGL releases, we got reports only in TGL_13 based.

**Comment by marc-hb** [comment]
> Validation teams has tried normal reboot cycles without copying any firmware & topologies: no issues reported.

Maybe this was just because the original firmware is not affected, only the new one.

**Comment by slawblauciak** [comment]
@marc-hb I was able to reproduce the issue as well. It only occurs after the FW binary is replaced.

**Comment by marc-hb** [comment]
> It only occurs after the FW binary is replaced.

With a different version? If not then it could be timing issue, has anyone tried to replace another, unrelated firmware file?

**Comment by slawblauciak** [comment]
> > It only occurs after the FW binary is replaced.
> 
> With a different version? If not then it could be a timing issue, has anyone tried to replace another, unrelated firmware file?
> 
> A delay can change the state of the DSP when the reboot starts.

No, the same version even.

**Comment by marc-hb** [comment]
OK then this must be a timing issue. Can you please try to copy similar and unrelated files in the same directories and see if it reproduces the same?
